pkg/trace/api/otlp.go

```diff
@@ -0,0 +1,506 @@
+	for i, e := range events {
+		if i > 0 {
+			str.WriteString(",")
+		}
+		var wrote bool
+		str.WriteString("{")
+		if v := e.TimeUnixNano; v != 0 {
+			str.WriteString(`"time_unix_nano":`)
+			str.WriteString(strconv.FormatUint(v, 10))
+			wrote = true
+		}
+		if v := e.Name; v != "" {
+			if wrote {
+				str.WriteString(",")
+			}
+			str.WriteString(`"name":"`)
+			str.WriteString(v)
+			str.WriteString(`"`)
+			wrote = true
+		}
+		if len(e.Attributes) > 0 {
+			if wrote {
+				str.WriteString(",")
+			}
+			str.WriteString(`"attributes":{`)
+			for j, kv := range e.Attributes {
+				if j > 0 {
+					str.WriteString(",")
+				}
+				str.WriteString(`"`)
+				str.WriteString(kv.Key)
+				str.WriteString(`":"`)
+				str.WriteString(anyValueString(kv.Value))
+				str.WriteString(`"`)
+			}
+			str.WriteString("}")
+			wrote = true
+		}
+		if v := e.DroppedAttributesCount; v != 0 {
+			if wrote {
+				str.WriteString(",")
+			}
+			str.WriteString(`"dropped_attributes_count":`)
+			str.WriteString(strconv.FormatUint(uint64(v), 10))
+		}
+		str.WriteString("}")
+	}
+	str.WriteString("]")
+	return str.String()
+}
+
+// convertSpan converts the span in to a Datadog span, and uses the rattr resource tags and the lib instrumentation
+// library attributes to further augment it.
+func convertSpan(rattr map[string]string, lib *otlppb.InstrumentationLibrary, in *otlppb.Span) *pb.Span {
+	name := spanKindName(in.Kind)
+	if lib.Name != "" {
+		name = lib.Name + "." + name
+	} else {
+		name = "opentelemetry." + name
+	}
+	span := &pb.Span{
+		Name:     name,
+		TraceID:  byteArrayToUint64(in.TraceId),
+		SpanID:   byteArrayToUint64(in.SpanId),
+		ParentID: byteArrayToUint64(in.ParentSpanId),
+		Start:    int64(in.StartTimeUnixNano),
+		Duration: int64(in.EndTimeUnixNano) - int64(in.StartTimeUnixNano),
+		Service:  rattr["service.name"],
+		Resource: in.Name,
+		Meta:     rattr,
+		Metrics: map[string]float64{
+			// auto-keep all incoming traces; it was already chosen as a keeper on
+			// the client side.
+			sampler.KeySamplingPriority: float64(sampler.PriorityAutoKeep),
+		},
+	}
+	if features.Has("otlp_original_ids") {
+		// keep original IDs
+		span.Meta["otlp_ids.trace"] = hex.EncodeToString(in.TraceId)
+		span.Meta["otlp_ids.span"] = hex.EncodeToString(in.SpanId)
+		span.Meta["otlp_ids.parent"] = hex.EncodeToString(in.ParentSpanId)
+	}
+	if _, ok := span.Meta["version"]; !ok {
+		if ver := rattr["service.version"]; ver != "" {
+			span.Meta["version"] = ver
+		}
+	}
+	if len(in.Events) > 0 {
+		span.Meta["events"] = marshalEvents(in.Events)
+	}
+	for _, kv := range in.Attributes {
+		switch v := kv.Value.Value.(type) {
+		case *otlppb.AnyValue_DoubleValue:
+			span.Metrics[kv.Key] = v.DoubleValue
+		case *otlppb.AnyValue_IntValue:
+			span.Metrics[kv.Key] = float64(v.IntValue)
+		default:
+			span.Meta[kv.Key] = anyValueString(kv.Value)
+		}
+	}
+	if _, ok := span.Meta["env"]; !ok {
+		if env := span.Meta["deployment.environment"]; env != "" {
```

Comment:
We should not use raw strings: semantic conventions are still unstable and it's better to import them from a package like https://pkg.go.dev/go.opentelemetry.io/otel@v0.20.0/semconv. The package might change in the future (see open-telemetry/opentelemetry-collector#2600) so we could do it later.

Another issue here is that at some point OpenTelemetry will introduce telemetry schemas (see doc [here](https://github.com/open-telemetry/oteps/blob/main/text/0152-telemetry-schemas.md)) which may change these constants. It's still early so I don't think this is a concern for now but it's something to keep in mind for designing this.

Reply:
I tried to avoid importing any OpenTelemetry stuff, but I guess this is a strong point. We should probably change this...

Reply:
Done.